### PR TITLE
Enforce a maximum queue depth

### PR DESCRIPTION
**Pull request by bbakerman**
#291 - The start of the ability to control the depth of queries

### CODE REVIEW

File: src/test/groovy/graphql/execution/ExecutionConstraintsTest.groovy [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+    ]
+
+
+    def 'Shows full data when the max depth is not exceeded'() {
+        given:
+
+        when:
+        def result = GraphQL
+                .newGraphQL(StarWarsSchema.starWarsSchema)
+                .maxQueryDepth(99)
```
Comment (bbakerman) [inline src/test/groovy/graphql/execution/ExecutionConstraintsTest.groovy]: Just to test that is not "unlimited" but rather not exceeded during the depth.  WQe have a case when its < 0 then its unlimited

File: src/main/java/graphql/execution/ExecutionStrategy.java [hunk 2]
```diff
@@ -21,7 +38,23 @@
...
+            public ExecutionResult call() throws Exception {
+                return resolveFieldImpl(executionConstraints, executionContext, parentType, source, fields);
+            }
+        });
+    }
+
+    private ExecutionResult resolveFieldImpl(ExecutionConstraints executionConstraints, ExecutionContext executionContext, GraphQLObjectType parentType, Object source, List<Field> fields) {
+        //
+        // we return null if the max depth has been exceeded to protect against large or malicious queries
+        if (executionConstraints.hasExceededMaxDepthConstraint()) {
```
Comment (bbakerman) [inline src/main/java/graphql/execution/ExecutionStrategy.java]: Yeah the library does not have a good way for a consumer to "inject" a strategy in for handling errors.

File: src/main/java/graphql/execution/ExecutionConstraints.java [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+public class ExecutionConstraints {
+    //
+    // future versions of this class could have "query complexity" etc. in it
+    // but for now we just have query depth
+    //
+    private final int maxQueryDepth;
+    private final AtomicInteger queryDepth;
+
+    private ExecutionConstraints(int maxQueryDepth) {
+        this.maxQueryDepth = maxQueryDepth;
```
Comment (Jimexist) [inline src/main/java/graphql/execution/ExecutionConstraints.java]: might be a good thing to assert that it is either 0 (unlimited) or positive?

**Comment by bbakerman** [comment]
This is not ready to be merged.  So please dont merge it.  

* I think the ExecutionConstraints should be built BEFORE being passed into the GraphQL object
* It probably should be an interface
* There is some challenges around depth here.  If we use a multi threaded execution strategy and we pick apart a spec like

   { 
      data {
        a,
        b,
        c
    }
}

and we spread a,b,c to threads each - are we at depth 1 or 4?

Perhaps we should calculate depth BEFORE executing it.  Eg run a tree count on the Document rather than execute it.

**Comment by IamCornholio** [comment]
> I think the ExecutionConstraints should be built BEFORE being passed into the GraphQL object
> 
> It probably should be an interface

I think these would go hand in hand. But do you mean built or defined? Could one define a constraint dependent on a graphql field that has not been "built" yet?

> There is some challenges around depth here. If we use a multi threaded execution strategy and we pick apart a spec like
> 
> {
> data {
> a,
> b,
> c
> }
> }
> 
> and we spread a,b,c to threads each - are we at depth 1 or 4?
> 
> Perhaps we should calculate depth BEFORE executing it. Eg run a tree count on the Document rather than execute it.

I haven't spent much time on thinking but I can't think of a useful constraint that can safely run multi-threaded on multiple nodes of a tree at once. Any constraint trying to constrain a tree in anyway through traversal would fall in this category. If I'm missing the obvious then perhaps you could introduce a flag and/or a class hierarchy of two types of constraints (single threaded and multi threaded execution constraints). 

You could also assume something safe and implement something that would act as a default. Since this is to become an extensible thing someone implementing graphql for their api would just implement what they need.

**Comment by kaqqao** [comment]
I believe Sangria also calculates the maximum possible depth (in case there are conditional parts) before even executing the query. This seems to be a decent approach.

**Pull request closed without merging**